substrate/primitives/core/src/address_uri.rs

```diff
@@ -0,0 +1,305 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Little util for parsing an address URI. Replaces regular expressions.
+
+/// A container for results of parsing the address uri string.
+///
+/// Intended to be equivalent of:
+/// `Regex::new(r"^(?P<phrase>[a-zA-Z0-9 ]+)?(?P<path>(//?[^/]+)*)(///(?P<password>.*))?$")`
+/// which also handles soft and hard derivation paths:
+/// `Regex::new(r"/(/?[^/]+)")`
+///
+/// Example:
+/// ```
+/// 	use sp_core::crypto::AddressUri;
+/// 	let manual_result = AddressUri::parse("hello world/s//h///pass");
+/// 	assert_eq!(
+/// 		manual_result.unwrap(),
+/// 		AddressUri { phrase: Some("hello world"), paths: vec!["s", "/h"], pass: Some("pass") }
+/// 	);
+/// ```
+#[derive(Debug, PartialEq)]
+pub struct AddressUri<'a> {
+	/// Phrase, hexadecimal string, or ss58-compatible string.
+	pub phrase: Option<&'a str>,
+	/// Key derivation paths, ordered as in input string,
+	pub paths: Vec<&'a str>,
+	/// Password.
+	pub pass: Option<&'a str>,
+}
+
+/// Errors that are possible during parsing the address URI.
+#[allow(missing_docs)]
+#[derive(Debug, thiserror::Error, PartialEq, Eq, Clone, Copy)]
+pub enum Error {
+	#[error("Invalid character in phrase")]
+	InvalidCharacterInPhrase,
+	#[error("Invalid character in password")]
+	InvalidCharacterInPass,
+	#[error("Invalid character in hard path")]
+	InvalidCharacterInHardPath,
+	#[error("Invalid character in soft path")]
+	InvalidCharacterInSoftPath,
+}
+
+fn extract_prefix<'a>(input: &mut &'a str, is_allowed: &dyn Fn(char) -> bool) -> Option<&'a str> {
+	let output = input.trim_start_matches(is_allowed);
+	let prefix_len = input.len() - output.len();
+	let prefix = if prefix_len > 0 { Some(&input[..prefix_len]) } else { None };
+	*input = output;
+	prefix
+}
+
+fn strip_prefix(input: &mut &str, prefix: &str) -> bool {
+	if let Some(stripped_input) = input.strip_prefix(prefix) {
+		*input = stripped_input;
+		true
+	} else {
+		false
+	}
+}
+
+impl<'a> AddressUri<'a> {
+	/// Parses the given string.
+	pub fn parse(mut input: &'a str) -> Result<Self, Error> {
+		let phrase = extract_prefix(&mut input, &|ch: char| {
+			ch.is_ascii_digit() || ch.is_ascii_alphabetic() || ch == ' '
+		});
+
+		let mut pass = None;
+		let mut paths = Vec::new();
+		while !input.is_empty() {
+			let unstripped_input = input;
+			if strip_prefix(&mut input, "///") {
+				pass = match extract_prefix(&mut input, &|ch: char| ch != '\n') {
+					Some(pass) => Some(pass),
+					None => Some(""),
+				};
+			} else if strip_prefix(&mut input, "//") {
+				let mut path = extract_prefix(&mut input, &|ch: char| ch != '/')
+					.ok_or(Error::InvalidCharacterInHardPath)?;
+				assert!(path.len() > 0);
+				// hard path shall contain leading '/', so take it from unstripped input.
+				path = &unstripped_input[1..path.len() + 2];
+				paths.push(path);
```

Comment:
The `path` here doesn't have to be mutable. (:

```suggestion
				let path = extract_prefix(&mut input, &|ch: char| ch != '/')
					.ok_or(Error::InvalidCharacterInHardPath)?;
				assert!(path.len() > 0);
				// hard path shall contain leading '/', so take it from unstripped input.
				paths.push(&unstripped_input[1..path.len() + 2]);
```